### ISSUE TITLE
Add source map support

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#376]
I don't know if this has been requested or not before, and would probably not be easy to implement, but I'll throw it out there: Source map support would be useful.  This would allow Duktape to map a position in the source being executed to its position in the original source, helpful for stack traces and such not only for minified code (its obvious application), but transcompiled code as well, e.g. CoffeeScript.

**Comment by svaarala** [comment]
It's listed as debugger future work, but it's OK to have an issue too: https://github.com/svaarala/duktape/blob/master/doc/debugger.rst#source-maps

**Comment by fatcerberus** [comment]
Tip: You might want to remove pause on error from future work. :)

**Comment by fatcerberus** [comment]
It occurs to me that this is not easily implementable at present because Duktape only tracks down to the line number level (pc2line).  For CoffeeScript which typically *adds* lines, this would be fine, but sourcemapping minified scripts will require more granularity, i.e. down to the character.

I was actually attempting to do this outside of Duktape, but unfortunately it's not really possible.  I can leverage `Duktape.errCreate()` to re-map `.fileName` and `.lineNumber` for errors, but that doesn't help for, e.g. debugger stepping.

For now what I'm doing is re-mapping the script filename (i.e. pass the source filename to `duk_pcompile()`) and just avoid minification when someone runs Cell (the minisphere compiler) requesting a source map.  Which is honestly probably the right solution anyway. :-)

**Comment by svaarala** [comment]
One place to apply source maps would be during compilation: at that point the exact line/offset information is available from the tokenizer, and the pc2line map could then reflect unminified source.

This is not a 100% solution because technically a single minified file can contain source code from multiple original source files and there's no way to reflect that in the compiled function output (i.e. `.fileName` is assumed to be a single value and not line dependent).

Adding lexer offset information into pc2line would be mostly quite simple: it's a bit-packed format for encoding line deltas, and roughly the same approach would work for column/offset information too (although the deltas will vary more widely).

**Comment by fatcerberus** [comment]
For me that second thing (multiple files minified into one .js file) isn't a problem because Cell works at the individual file level (1 asset = 1 file) but that's a good point in the general case.

When I actually have some free time I might look into implementing this.  Another thing we would have to figure out is how to get the source map itself into Duktape, i.e. What the API would be like.

**Comment by fatcerberus** [comment]
I imagine that bit-packed format works quite well for line numbers: In 95% of cases you differ by either 0 or 1 line, so you only need a single bit.

**Comment by svaarala** [comment]
IIRC it was roughly 3 bits / opcode for usual code; an opcode is 32 bits so it's roughly 10% of bytecode size.

**Comment by svaarala** [comment]
That pc2offset information probably doesn't need any random access like pc2line does, so it doesn't need a rough index like the one pc2line has (there's a bitstream starting point every 64 opcode indices or something like that). Basically it should suffice to deliver that pc2offset information to the debug client - and then let the debug client use pc2line, pc2offset, and the source map to figure out the rest?

**Comment by svaarala** [comment]
It would probably be good that the debug command to request line/offset information would return unpacked line/offset information rather than the raw pc2line format - it's an internal format and it'd be best to hide it from the debug client because it may change at any time.

**Comment by fatcerberus** [comment]
Agreed, best to hide internal details as much as possible, especially in the experimental stage.  No use locking yourself into a format early.

**Comment by fatcerberus** [comment]
The problem with letting the client do it, at least in my case, is that the source map is packaged with the game in the compiled .spk file, so only the target (the game engine) can access it.  Duktape also needs access to the map for, e.g. Error augmentation.

**Comment by svaarala** [comment]
With respect to the pc2line format, I don't see a reason to expose it even if it were relatively unchanged (as it actually is, I haven't touched it in ages). One reason is that there may be a legitimate reason to even have multiple formats. For example, if memory is not a particular issue, a simple array of pc-to-line mappings is much faster and simpler.

**Comment by svaarala** [comment]
The default error augmentation processing actually doesn't set `.fileName` or `.lineNumber`, it simply collects the tracedata which is an array referencing the relevant function objects and their PC values. An inherited getter than translates a `.fileName` property lookup (for example) into a value based on the tracedata.

Since that getter has access to tracedata, it also has access to the function and its pc2line and pc2offset data. If a function further had a source map reference, the getter could then complete the process. For this process it'd be easiest if the source map was yet another normal or internal property.

Conceptually anyway; I'm not sure what the best formats would be and if pc2line/pc2offset should be a combined bit stream etc as they'd be sometimes looked up together. Further, on low memory targets there are significant constraints in increasing function footprint so it may be more appropriate for them to handle source maps another way.

**Comment by fatcerberus** [comment]
I suspected that's what the `.fileName` and `.lineNumber` getter was doing, reading from the stack trace.  As for footprint, I would definitely want to make source maps an optional feature hidden behind a feature option (e.g. DUK_USE_SOURCE_MAP) because I imagine it would add a least a couple KB of code size.

Anyway, I'll see if I can draft something up.

**Comment by fatcerberus** [comment]
Hm, revisiting this now, I realize exposing line and column through either the C API, debug protocol or both would be sufficient for my use case and probably most others.  This alleviates the need for Duktape to settle on a source map format and leaves it up to the application.

**Comment by fatcerberus** [comment]
Coming back to this, my point above (mostly) stands.

I don't really expect debugging of minified scripts to work well in any case, and minisphere has a build system (the Cell compiler) as part of its toolchain, so builds enabled for can just skip the minification.

However, assuming I would want to debug a minified game, the most natural way for that work (to me) would be to step at the bytecode level based on a source map, with Duktape providing line and column information which could be used--by the client--to map back to the original source.  The basic requirements of Duktape would be:
* A mechanism for stepping through bytecode (one opcode at a time).  I imagine this would be fairly easy to implement with negligible impact on footprint/complexity, since we already have the stepping logic, it would just need to be adapted to work in the `pc` domain instead of the line number domain.  This would also have other use cases, e.g. a "disassembling" debugger.
* The ability for Duktape to map a `pc` value to both a line number and (perhaps approximate) column number.  The client could then consult the source map and decide when to stop sending step commands.  Even outside of debugging, this would be useful for, e.g. error reports--the source map lookup can then be done offline to find the location of the bug.

Extras that would be nice to have but not strictly required would be:
* The ability to step an arbitrary number of opcodes at a time (to avoid transport overhead).  This would be more complex since a multi-step operation would need to take into account breakpoints, etc. that the hardcoded single-step operation can safely ignore.  GDB for example allows multi-step but stops if it hits a signal or breakpoint first.

The important thing is, Duktape wouldn't need to maintain or parse a source map on the target.  As long as the above primitives are in place the bulk of the work can be done by the client.

**Comment by svaarala** [comment]
@fatcerberus That's a pretty clean approach in principle but tracking column information gets a bit messy (it's much less compressible than pc2line information for example). And also setting breakpoints for either PC or line+column has a lot of corner cases.

**Comment by svaarala** [comment]
By the way, the current "single step" operation doesn't do any shortcuts such as skipping breakpoints or such. It will just setup the "interrupt counter" to 1 so that we'll execute one bytecode instruction (not line!) and we end up back in the executor interrupt routine to determine if the line number has changed, i.e. we've completed a "step line" operation.

But that check runs for every opcode executed, so supporting a "step one instruction" would be trivial as such.

**Comment by fatcerberus** [comment]
I figured as much, having read about "checked execution" in the debugger docs (yes, I *did* read them!).  What would be more complex (I assume) is a multi-step function, since it would need to bail out if it hit a breakpoint.  Although I guess that could be covered by the existing logic too.

I do remember seeing a comment at one point to the effect of "we want to avoiding trigger the same breakpoint while stepping to the next line" though.

**Comment by svaarala** [comment]
Yeah, but that logic would be in the interrupt handler (driven by the interrupt counter). It can be slow because when we're *not* stepping the executor will only track the interrupt counter and call into the interrupt very very infrequently.

So in short, making that logic more complex for stepping is not a performance issue, and checking for any conditions for every instruction executed is quite OK (and already done).

**Comment by svaarala** [comment]
That stepping and breakpoint comment is related to how the semantics of "step a line" are defined: we want to trigger when the step is complete, i.e. we transition from "current line matches breakpoint" to "current line doesn't match breakpoint", and the code is tracking that.

But it's all in the interrupt handler, which executes for every instruction, so it's quite easy to muck around with. The interrupt counter is the only mechanism affecting the core executor, and it's always configured to 1 when stepping so that any conditions can be evaluated for every instruction. There's no support for e.g. understanding a line transition in the core executor which is very nice IMO.

**Comment by fatcerberus** [comment]
I think I'll look into the interrupt handler again and perhaps start hacking on it. :)  I'm already somewhat familiar with it since I needed to deal with it to implement Throw notify.

**Comment by svaarala** [comment]
(That interrupt counter is quite annoying to track correctly but it's critical that there is only one counter for interrupt execution once in a while - any change to that is an immediate performance hit in the order of 10-30%. I think the current division works pretty well :-)

**Comment by svaarala** [comment]
Adding support for debugger features may require quite big changes in the debugger state management for breakpoints, etc, but I'm very much hoping none of it will impact the bytecode executor which is critical to performance and correctness. I have pretty high confidence it will also turn out that way in practice :)

**Comment by fatcerberus** [comment]
Yes, it's quite a critical hot path and very easy to introduce a performance regression--as I found out firsthand.

**Comment by fatcerberus** [comment]
It seems there's a de facto standard format for source maps these days:
https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit

I know the V3 format is used by at least CoffeeScript and TypeScript.  Not sure about Babel.

**Comment by andrewemmons** [comment]
Hello! Just wanted to request this feature as well, or similar way where we can debug code that has run through Babel.

**Comment by denisdemaisbr** [comment]
Well then of what use is the source maps?
